---
title: "GALA_olink"
author: "Helene BÃ¦k Juel"
date: "8/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Load packages
source(here::here("R/package-loading.R"))
```

## Load full dataset
Can be skipped if olink3/4 loaded directly.
```{r load-data, include=FALSE}
# Load data from raw NPX data
olink <- OlinkAnalyze::read_NPX(here::here("data-raw/20202249_Juel_NPX.xlsx"))

# Fix issue with TIPS naming
# replace typo TIEP80EC	--> TIPS80EC
olink2 <- olink %>%
    mutate(SampleID = sub("TIEP80EC", "TIPS80EC", SampleID))
# replace typo TIPSKC/93 --> TIPS93
olink2 <- olink2 %>%
    mutate(SampleID = sub("TIPSKC/93", "TIPS93", SampleID))

# Add columns for cohort and pt_ID to be able to extract each cohort
olink2 <- olink2 %>%
    mutate(cohort = sub("[0-9]{1,4}.*$", "", SampleID)) %>% # substitute 1-4 numbers and everything after this (.*$) with nothing
    mutate(cohort = sub("_.*$", "", cohort)) # substitute anything after a _ with nothing to remove long control names and just keep "CONTROL"

# change cytokine names
# first remove all -
# then change spaces to _
# then rename alpha/beta/gamma to a/b/g
olink3 <- olink2 %>%
    mutate(Assay = gsub("-", "", Assay)) %>%
    mutate(Assay = gsub(" ", "_", Assay)) %>%
    mutate(Assay = gsub("alpha", "a", Assay)) %>%
    mutate(Assay = gsub("beta", "b", Assay)) %>%
    mutate(Assay = gsub("gamma", "g", Assay))

# change values < LOD of that assay to 50% LOD
olink4 <- olink3 %>%
    mutate(corr_NPX = if_else(LOD > NPX, LOD/2, NPX))
# TODO: consider whether this is needed - and if so, overwrite NPX (used in Olinkanalyze package)
```

# Wrangling and QC

## Filter GALA samples
```{r QC1}
# Load dataset if not running "Load full dataset" chunk above
# load(here::here("data/olink4.rda"))

GALA_olink <- olink3 %>% # olink4 has values below LoD substituted with 50% LoD. Olink3 uses the measured NPX values
    filter(cohort == "ALD" | cohort == "HP") %>%
    mutate(sample_type = sub("^.+[0-9]{1,3}", "", SampleID)) # duplicate sample marked as "-d"

# Remove duplicate samples
GALA_olink <- GALA_olink %>%
    filter(sample_type != "-d")  # goes from 50692 to 50416 obs

GALA_olink$sample_type <- NULL # remove sample_type column

```


## Visualize potential outliers by IQR vs. sample median
```{r QC_plot}
olink_qc_plot(GALA_olink)
```

Some samples have a QC warning and 2 appear to be outliers

## Further QC
```{r QC2}
# remove samples with QC warning
GALA_olink <- GALA_olink %>%
    filter(QC_Warning == "Pass") # goes from 50416 to 49404 obs

# Remove extra 0's in SampleID HP004-HP096 to be able to merge with phenotype data
GALA_olink <- GALA_olink %>%
  mutate(SampleID = sub("^HP[0]{1,2}", "HP", SampleID))

# Remove cohort column before merging with phenotype data
GALA_olink$cohort <- NULL

```

## Load and wrangle phenotype data
Can be skipped if merged dataset "GALA" loaded directly.
```{r phenotype1, warning=FALSE}
# TODO: Change to new QC'ed phenotype dataset
GALA_pheno <- read_excel(here::here("data-raw/ALD_HP_outcome_HBJ.xlsx"))
str(GALA_pheno)

# Change data types
GALA_pheno2 <- GALA_pheno %>% 
  mutate(kleiner_numeric = kleiner, 
         steatosis_numeric = nas_steatosis, 
         inflam_numeric = nas_inflam) %>% #Add numeric columns (before parent column is converted to factor) 
  mutate(across(c(cpa, meld, elf, te, # liver scores
                  packyears, 
                  height, weight, bmi, waist, hip, whr, # anthropometry
                  hr, map, sbp, dbp, # heart
                  alt, ast, alk, bili, crp, ggt, # liver enzymes
                  chol, hdl, ldl, trigly, # lipids
                  iga, igg, igm, leu,  # antibodies
                  glc, hba1c, insulin, homair, cpeptid, # glucose metabolism
                  proc3, 
                  days_to_LRE, days_to_hospInf, days_to_mort, 
                  kleiner_numeric, steatosis_numeric, inflam_numeric), as.numeric)) %>% # Change data columns to numeric
    mutate(te_fibrosis_6 = if_else(te <= 6, "low", "high"),
           te_fibrosis_8 = if_else(te <= 8, "low", "high"), # Add columns for fibrosis level high or low based on TE
           steatosis_binary = if_else(steatosis_numeric < 2, "low", "high")) %>% 
# TODO: also add binary inflammation and steatosis
  mutate(across(c(cohort, gender, 
                  kleiner, fibrosis, starts_with("nas"), inflam, # liver scores
                  abstinent, overuse, alcoholyears, abstinenceyears, # alcohol
                  liverrelated_event, hospInfection, allMortality, # outcomes
                  te_fibrosis_6, te_fibrosis_8, steatosis_binary,
                  excess_drink_followup), factor)) %>% # Change data columns to factor
# TODO: order factors alcoholyears and abstinenceyears
  mutate(kleiner_numeric = if_else(is.na(kleiner_numeric), 0.5, kleiner_numeric)) # assign NA to 0.5, since those without a kleiner score are known to have no or low fibrosis

# Add more endpoint columns 
GALA_pheno3 <- GALA_pheno2 %>%  
  mutate(days_to_hospInf2 = dplyr::if_else(is.na(days_to_hospInf), ifelse(allMortality == "no", 2*max(days_to_hospInf, na.rm = T), NA), days_to_hospInf), # If the patient did not have a hospInf (n=321), assign 2x max(days_to_hospInf) to pts who did not die and NA to those who died without having an infection (n=31). Those that were not followed up (n=155) retain NA. 
         #Note that dplyr::if_else() requires T and F values to be of the same type, hence you cannot assign NA together with a numeric vector. baseR ifelse() does not have this constraint.
         days_to_LRE2 = dplyr::if_else(is.na(days_to_LRE), ifelse(allMortality == "no", 2*max(days_to_LRE, na.rm = T), NA), days_to_LRE), # no LRE n=378, no follow-up n=155
         days_to_mort2 = dplyr::if_else(allMortality == "no", 2*max(days_to_mort, na.rm = T), ifelse(allMortality == "yes", days_to_mort, NA)), # no death n=386, no follow-up n=155
         hospInf_1yr = as.factor(ifelse(days_to_hospInf2 == "NA", NA, if_else(days_to_hospInf2 < 366, "yes", "no"))), # NA = not followed up or died without having an infection (n=186); yes = the pt was followed up AND had an infection in the first year (n=39); no = the pt was followed up AND DID NOT have an infection in the first year (n=392)
         LRE_1yr = as.factor(ifelse(days_to_LRE2 == "NA", NA, if_else(days_to_LRE2 < 366, "yes", "no"))), # n=33
         mort_1yr = as.factor(ifelse(days_to_mort2 == "NA", NA, if_else(days_to_mort2 < 366, "yes", "no"))), # n=11
         any_event_tot = as.factor(if_else(c(hospInfection == "yes" | liverrelated_event == "yes" | allMortality == "yes"), "yes", ifelse(liverrelated_event == "NA", NA, "no"))), # n=184
         any_event_1yr = as.factor(if_else(c(hospInf_1yr == "yes" | LRE_1yr == "yes" | mort_1yr == "yes"), "yes", ifelse(c(liverrelated_event == "NA" | hospInfection == "NA" | allMortality == "NA"), NA, "no")))) # n=65
# summary(GALA_pheno3[, 54:74])

# Change column name from CBMR_ID to SampleID to allow merging with olink data
GALA_pheno3 <- GALA_pheno3 %>%
  rename(SampleID = CBMR_ID)

# Change order of factors, if needed
GALA_pheno3$cohort <- ordered(GALA_pheno3$cohort, levels = c("HP", "ALD"))

# str(GALA_pheno3)

# save dataset in data folder
# usethis::use_data(GALA_pheno3, overwrite = T)
```


## Match and merge Olink with phenotype data
Can be skipped if merged dataset "GALA" loaded directly.
```{r merge1}
# Match and merge Olink with phenotype data
GALA0 <- merge(GALA_olink, GALA_pheno3, by = "SampleID") # goes from 49404 to 49312 obs from GALA_olink = 1 ppt!

# Find missing sample
print(setdiff(unique(GALA_olink$SampleID), GALA_pheno3$SampleID)) # ALD1302

```
TODO: replace phenotype dataset with final QC'd version
TODO: add phenotype data for ALD1302

## Add genotype SNP and PRS datasets
Can be skipped if merged dataset "GALA" loaded directly.
```{r phenotype2}
# Add PRS and SNP data
ALD_gen <- read_excel(here::here("data-raw/genALD_HBJ.xlsx"), sheet = "combined") #489 obs
HP_gen <- read_excel(here::here("data-raw/genHP_HBJ.xlsx"), sheet = "combined") #136 obs
GALA_gen <- rbind(ALD_gen, HP_gen) #625 obs
# Change column name from CBMR_ID to SampleID to allow merging with olink data
GALA_gen <- GALA_gen %>%
  rename(SampleID = CBMR_ID)

# save dataset in data folder
# usethis::use_data(GALA_gen, overwrite = T)
```

## Merge genotypes with Olink+phenotype data
Can be skipped if merged dataset "GALA" loaded directly.
```{r merge2}
GALA <- merge(GALA0, GALA_gen, by = "SampleID") # goes from 49312 to 48852 obs from GALA0 = 5 ppts
print(setdiff(unique(GALA0$SampleID), GALA_gen$SampleID)) # ALD2265, ALD2332, ALD2547, ALD2558, ALD402

# str(GALA[,c(74:108)]) # check

# save dataset in data folder
usethis::use_data(GALA, overwrite = T)
```
TODO: investigate why we are missing genotypes for these 5

## Load merged dataset
```{r load_merged}
# Load dataset if not running the loading and merging chunks above
load(here::here("data/GALA.rda"))
```


# Results

## Plot all cytokines vs Kleiner score
Not included in current knit (include=FALSE).
```{r plot_all_kleiner, warning=FALSE, include=FALSE}
# Olink ID lists for looping
cytokines <- levels(as.factor(olink4$Assay)) # protein abbreviations
cytokine_IDs <- levels(as.factor(olink4$OlinkID)) #Olink IDs

# Loop over all cytokines
allcytokines_kleiner <- list()

for(i in cytokines[1:92]) {
    ploti <- GALA %>%
        filter(Assay == i) %>%
        ggplot(aes(x = kleiner, y = NPX, color = cohort)) +
        geom_violin(draw_quantiles = c(0.5)) +
        geom_jitter() +
        ggtitle(paste(i))

    allcytokines_kleiner[[i]] <- ploti
    print(ploti)
}

# TODO: fix scatterplot in NA column and place NA first

# To save plotgrid as pdf
# pdf(here::here("doc/images/GALA_allcytokines_kleiner.pdf"), height = 60, width = 40) # height = 5 for each row of 8 -> 60 for 12x8
# do.call('grid.arrange', c(allcytokines_kleiner, ncol=8))
# dev.off()

```

## Run ANOVA for Kleiner score

Plots are ordered by ascending p-val.
```{r ANOVA_kleiner}
# count n
# summary(as.factor(GALA$kleiner)) # F0: 32, F1: 114, F2: 88, F3: 20, F4: 56, NA: 226. Sum !=NA: 310

# run ANOVA
anova_GALA_kleiner <- olink_anova(GALA, variable = "kleiner")

# Extract list of significant cytokines
sign_cytokines <- anova_GALA_kleiner %>%
    filter(Adjusted_pval < 0.05/92) %>% # Bonferroni = 92 tests
    pull(Assay)
print(sign_cytokines)
```
72 of the 92 cytokines have p<0.05 after Bonferroni correction for multiple testing.

## Plot significant cytokines (Kleiner)
Not included in current knit (include=FALSE).
```{r plot_sign_kleiner, warning=FALSE, include=FALSE}
# Prepare list for graphs
signcytokines_kleiner <- list()

for(i in sign_cytokines[1:72]) {
    ploti <- GALA %>%
        # filter(!is.na(as.numeric(kleiner))) %>%
        filter(Assay == i) %>%
        ggplot(aes(x = kleiner, y = NPX, color = cohort)) +
        geom_violin(draw_quantiles = c(0.5)) +
        # geom_jitter(position = position_jitterdodge(), size = 0.5) + # jitterdodge is throwing an error, investigate!
        geom_jitter(size = 0.5) +
        ggtitle(paste(i))

    signcytokines_kleiner[[i]] <- ploti
    print(ploti)
}

# To save plotgrid as pdf
# pdf(here::here("doc/images/GALA_signcytokines_kleiner.pdf"), height = 45, width = 40) # height = 5 for each row of 8 -> 45 for 9x8
# do.call('grid.arrange', c(signcytokines_kleiner, ncol=8))
# dev.off()
```
Observations from plots:

- Most cytokines (except SCF) increase with increasing Kleiner score
- Some cytokines appear to increase already from Kleiner 1 to 2 (HGF, IL8, CCL19, CDCP1, LAP-TGFb, others)
- Some cytokines appear to increase only with high fibrosis (SLAMF1, uPA, IL15RA, PDL1, CCL28, CD8A TNFb, others)
- Need to consider how to analyse cytokines with many observations below LoD (e.g. IL6, FGF23, GDNF, SIRT2, ST1A1, MCP3, IL17A, ARTN)
- FGF21 appears to have a bell-shaped curve, with maximum at Kleiner 2. Interestingly, a significant increase in FGF21 was observed 24h(? TP3) after alcohol bolus in AlcoChallenge, in both liver vein and peripheral blood. No other cytokines showed significant increase (HGF and NRTN decreased).

Other interesting findings to consider:
- IL6, OSM, CXCL10, CCL19, MCP4 all significantly higher in liver vein compared with peripheral blood in AlcoChallenge study --> being produced by the liver? 

## Linear model for TE
```{r lm_te}
# Run linear mixed model
# GALA$te <- as.numeric(GALA$te)

# count n
# summary(GALA$te) # 18 NA

lm_GALA_te <- olink_lmer(GALA, 
           variable = "te", 
           outcome = "NPX",
           random = "cohort")

# Extract list of significant cytokines
sign_cytokines <- lm_GALA_te %>%
    filter(Adjusted_pval < 0.05/92) %>% # Bonferroni = 92
    pull(Assay)
print(sign_cytokines) # 57 cytokines significant (54 when using corr_NPX)
```
57 cytokines have p<0.05 after Bonferroni correction for multiple testing.

## Plot cytokines with significant linear correlation with TE
```{r plot_sign_TE, warning=FALSE}
# Prepare list for graphs
signcytokines_te <- list()

for(i in sign_cytokines[1:54]) {
    ploti <- GALA %>%
        filter(Assay == i) %>%
        ggplot(aes(x = as.numeric(te), y = corr_NPX)) +
        geom_point(aes(color = cohort)) +
        geom_smooth(method = lm, color = "black") +
        scale_x_log10() +
        ggtitle(paste(i))

    signcytokines_te[[i]] <- ploti
    print(ploti)
}

# To save plotgrid as pdf
# pdf(here::here("doc/images/GALA_signcytokines_te.pdf"), height = 35, width = 40) # height = 5 for each row of 8 -> 35 for 7x8
# do.call('grid.arrange', c(signcytokines_te, ncol=8))
# dev.off()
```

- In addition to SCF, TRAIL, CXCL5, CCL23 also show negative correlation with TE

## Cytokine functions of top correlations
- HGF hepatocyte growth factor regulates cell growth, cell motility and morphogenesis in numerous cell and tissue types (https://www.ncbi.nlm.nih.gov/gene/3082). Higher levels of HGF have been associated with the presence of peripheral arterial disease (doi: 10.1177/0003319720912935).
- IL8 AKA CXCL8 is a major mediator of the inflammatory response. IL-8 is secreted by mononuclear macrophages, neutrophils, eosinophils, T lymphocytes, epithelial cells, and fibroblasts. It functions as a chemotactic factor by guiding the neutrophils to the site of infection. Bacterial and viral products rapidly induce IL-8 expression. IL-8 also participates with other cytokines in the proinflammatory signaling cascade and plays a role in systemic inflammatory response syndrome (SIRS). The binding of IL-8 to one of its receptors (IL-8RB/CXCR2) increases the permeability of blood vessels and increasing levels of IL-8 are positively correlated with increased severity of multiple disease outcomes (e.g., sepsis) (https://www.ncbi.nlm.nih.gov/gene/3576). 
- IL6 functions in inflammation and the maturation of B cells. In addition, the encoded protein has been shown to be an endogenous pyrogen capable of inducing fever in people with autoimmune diseases or infections. The protein is primarily produced at sites of acute and chronic inflammation, where it is secreted into the serum and induces a transcriptional inflammatory response through interleukin 6 receptor, alpha. The functioning of this gene is implicated in a wide variety of inflammation-associated disease states, including suspectibility to diabetes mellitus and systemic juvenile rheumatoid arthritis (https://www.ncbi.nlm.nih.gov/gene/3569).
- uPA urokinase plasminogen activator (PLAU) is a secreted serine protease that converts plasminogen to plasmin (https://www.ncbi.nlm.nih.gov/gene/5328). 
- FGF21 fibroblast growth factor 21, (also FGF23). FGF family members possess broad mitogenic and cell survival activities and are involved in a variety of biological processes. This protein is a secreted endocrine factor that functions as a major metabolic regulator. The encoded protein stimulates the uptake of glucose in adipose tissue (https://www.ncbi.nlm.nih.gov/gene/26291). Exclusively produced in liver tissue. Appears to be regulated in several diseases (kidney, sarcopenia, aneurysms, DM2) 
- SLAMF1 signaling lymphocytic activation molecule family member 1 is of quite unknown function (https://www.ncbi.nlm.nih.gov/gene/6504).
- CCL20 The CC cytokines are proteins characterized by two adjacent cysteines. The protein encoded by this gene displays chemotactic activity for lymphocytes and can repress proliferation of myeloid progenitors (https://www.ncbi.nlm.nih.gov/gene/6364). 
- LIFR Leukemia inhibitory factor receptor belongs to the type I cytokine receptor family. This protein combines with a high-affinity converter subunit, gp130, to form a receptor complex that mediates the action of the leukemia inhibitory factor, a polyfunctional cytokine that is involved in cellular differentiation, proliferation and survival in the adult and the embryo. 
- SCF (negative correlation) stem cell factor = KITLG? This ligand is a pleiotropic factor that acts in utero in germ cell and neural cell development, and hematopoiesis, all believed to reflect a role in cell migration. In adults, it functions pleiotropically, while mostly noted for its continued requirement in hematopoiesis (https://www.ncbi.nlm.nih.gov/gene/4254). 


## Suggestions for moving on
1. Decide on outcome to assess/predict
    + Split into high vs low fibrosis (Kleiner score 2-4 vs 0-1, with TE<6kPA included in the low group)
    + Use Kleiner score (also add TE<6kPa to F0-1)
    + Use TE
    + Use Inflammation
    + Use outcome data (presence of liver-related event, presence of infection)
2. Run PCA (using decided outcome)
3. Generate machine learning model to predict decided outcome using 2-layer cross-validation
4. Plot heatmap showing which cytokines are correlated, then choose a few for further analysis
5. Volcano plot for all significant cytokines

### The following packages and versions were used to create this report
```{r sessioninfo}
sessionInfo()
```


### This document was knitted on
```{r knitdate}
Sys.Date()
```

